p2p/lightchain.py

```diff
@@ -245,35 +263,105 @@ def peer_block_height(peer: LESPeer) -> int:
+    async def get_block_header_by_hash(self, block_hash: Hash32) -> BlockHeader:
         peer = await self.get_best_peer()
-        self.logger.debug("Fetching header %s from %s", encode_hex(block_hash), peer)
-        return await peer.get_block_header_by_hash(block_hash, self.cancel_token)
+        return await self._get_block_header_by_hash(peer, block_hash)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_block_body_by_hash(self, block_hash: Hash32) -> BaseBlock:
         peer = await self.get_best_peer()
         self.logger.debug("Fetching block %s from %s", encode_hex(block_hash), peer)
-        return await peer.get_block_by_hash(block_hash, self.cancel_token)
+        request_id = gen_request_id()
+        peer.sub_proto.send_get_block_bodies([block_hash], request_id)
+        reply = await self._wait_for_reply(request_id)
+        if not reply['bodies']:
+            raise BlockNotFound("Peer {} has no block with hash {}".format(peer, block_hash))
+        return reply['bodies'][0]
 
     # TODO add a get_receipts() method to BaseChain API, and dispatch to this, as needed
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_receipts(self, block_hash: Hash32) -> List[Receipt]:
         peer = await self.get_best_peer()
         self.logger.debug("Fetching %s receipts from %s", encode_hex(block_hash), peer)
-        return await peer.get_receipts(block_hash, self.cancel_token)
+        request_id = gen_request_id()
+        peer.sub_proto.send_get_receipts(block_hash, request_id)
+        reply = await self._wait_for_reply(request_id)
+        if not reply['receipts']:
+            raise BlockNotFound("No block with hash {} found".format(block_hash))
+        return reply['receipts'][0]
 
     # TODO implement AccountDB exceptions that provide the info needed to
     # request accounts and code (and storage?)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_account(self, block_hash: Hash32, address: Address) -> Account:
         peer = await self.get_best_peer()
-        return await peer.get_account(block_hash, address, self.cancel_token)
+        key = keccak(address)
+        proof = await self._get_proof(peer, block_hash, account_key=b'', key=key)
+        header = await self._get_block_header_by_hash(peer, block_hash)
+        rlp_account = HexaryTrie.get_from_proof(header.state_root, key, proof)
+        return rlp.decode(rlp_account, sedes=Account)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_contract_code(self, block_hash: Hash32, key: bytes) -> bytes:
         peer = await self.get_best_peer()
-        return await peer.get_contract_code(block_hash, key, self.cancel_token)
+        request_id = gen_request_id()
+        peer.sub_proto.send_get_contract_code(block_hash, key, request_id)
+        reply = await self._wait_for_reply(request_id)
+        if not reply['codes']:
```

Comment:
It seems like we should be looking up the `account.code_hash` and verifying that the returned code is indeed the code which hashes to the expected `code_hash`?

Reply:
Well, this is the light client, so we can't assume we'll have the account in our db, right?

Reply:
Right right right, context... 👍

Reply:
Um, but wait, we could still fetch the account which IIRC comes back as a merkle proof which is verifiable and then we can verify the code against the code hash.

Reply:
Yeah, I guess so. Filed #849